// reg-cli/dist/process-adaptor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _child_process = require("child_process");
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// $FlowIgnore
class ProcessAdaptor {
  constructor(emitter) {
    this._process = (0, _child_process.fork)(_path.default.resolve(__dirname, './diff.js'));
    this._isRunning = false;
    this._emitter = emitter;
  }
  isRunning() {
    return this._isRunning;
  }
  run(params) {
    return new Promise((resolve, reject) => {
      this._isRunning = true;
      if (!this._process || !this._process.send) resolve();
      this._process.send(params);
      this._process.once('message', result => {
        this._isRunning = false;
        this._emitter.emit('compare', {
          type: result.passed ? 'pass' : 'fail',
          path: result.image
        });
        resolve(result);
      });
    });
  }
  close() {
    if (!this._process || !this._process.kill) return;
    this._process.kill();
  }
}
exports.default = ProcessAdaptor;
